File: java/library/src/main/java/com/v2soft/styxlib/Connection.java
package com.v2soft.styxlib;

import com.v2soft.styxlib.handlers.RMessagesProcessor;
import com.v2soft.styxlib.handlers.TMessageTransmitter;
import com.v2soft.styxlib.exceptions.StyxException;
import com.v2soft.styxlib.messages.StyxRAttachMessage;
import com.v2soft.styxlib.messages.StyxRAuthMessage;
import com.v2soft.styxlib.messages.StyxRVersionMessage;
import com.v2soft.styxlib.messages.StyxTAttachMessage;
import com.v2soft.styxlib.messages.StyxTAuthMessage;
import com.v2soft.styxlib.messages.StyxTVersionMessage;
import com.v2soft.styxlib.messages.base.StyxMessage;
import com.v2soft.styxlib.messages.base.StyxTMessageFID;
import com.v2soft.styxlib.messages.base.enums.MessageType;
import com.v2soft.styxlib.messages.base.structs.StyxQID;
import com.v2soft.styxlib.server.ClientDetails;
import com.v2soft.styxlib.server.IChannelDriver;
import com.v2soft.styxlib.server.IMessageTransmitter;
import com.v2soft.styxlib.library.types.ConnectionDetails;
import com.v2soft.styxlib.library.types.Credentials;

import java.io.Closeable;
import java.io.IOException;
import java.util.concurrent.TimeoutException;

/**
 * Styx client conection
 *
 * @author V.Shcryabets <a>vshcryabets@gmail.com</a>
 */
public class Connection
        implements Closeable, IClient {
    //---------------------------------------------------------------------------
    // Constants
    //---------------------------------------------------------------------------
    public static final String PROTOCOL = "9P2000";
    public static final int DEFAULT_TIMEOUT = 10000;
    private static final int DEFAULT_IO_SIZE = 8192;
    //---------------------------------------------------------------------------
    // Class fields
    //---------------------------------------------------------------------------
    private StyxFile mRoot;
    protected Credentials mCredentials;
    private String mMountPoint;
    private int mTimeout = DEFAULT_TIMEOUT;
    private boolean isConnected, isAttached;
    private TMessageTransmitter mTransmitter;
    private long mAuthFID = StyxMessage.NOFID;
    private StyxQID mAuthQID;
    private StyxQID mQID;
    private long mFID = StyxMessage.NOFID;

    protected ClientDetails mRecepient;
    private IChannelDriver mDriver;
    protected ConnectionDetails mDetails;
    protected RMessagesProcessor mAnswerProcessor;
    protected boolean isAutoStartDriver = false;
    protected boolean shouldCloseAnswerProcessor = false;
    protected boolean shouldCloseTransmitter = false;

    public Connection() {
        this(new Credentials(null, null));
    }

    public Connection(Credentials credentials) {
        this(credentials, null);
    }

    public Connection(Credentials credentials, IChannelDriver driver) {
        this(credentials, driver, null, null, null);
    }

    public Connection(Credentials credentials,
                      IChannelDriver driver,
                      RMessagesProcessor answerProcessor,
                      TMessageTransmitter transmitter,
                      ClientDetails recepient) {
        mTransmitter = transmitter;
        mRecepient = recepient;
        mAnswerProcessor = answerProcessor;
        mDriver = driver;
        mDetails = new ConnectionDetails(getProtocol(), getIOBufSize());
        mCredentials = credentials;
        isConnected = false;
    }
    /**
     * Connect to server with specified parameters
     *
     * @param driver      channel driver
     * @param credentials user credentials
     * @return true if connected
     * @throws IOException
     * @throws StyxException
     * @throws TimeoutException
     */
    public boolean connect(IChannelDriver driver, Credentials credentials)
            throws IOException, StyxException, InterruptedException, TimeoutException {
        if ( mAnswerProcessor == null ) {
            mAnswerProcessor = new RMessagesProcessor("RH" + driver.toString());
            shouldCloseAnswerProcessor = true;
        }
        if ( mTransmitter == null ) {
            mTransmitter = new TMessageTransmitter(mTransmitterListener);
            shouldCloseTransmitter = true;
        }

        if (!driver.isStarted()) {
            driver.start(getIOBufSize());
            isAutoStartDriver = true;
        }

        if (mRecepient == null) {
            // get first client from driver
            mRecepient = driver.getClients().iterator().next();
        }

        return this.connect(driver, credentials, mAnswerProcessor, mTransmitter, mRecepient);
    }

    /**
     * Connect to server with specified parameters
     *
     * @param driver      channel driver
     * @param credentials user credentials
     * @return true if connected
     * @throws IOException
     * @throws StyxException
     * @throws TimeoutException
     */
    public boolean connect(IChannelDriver driver, Credentials credentials, RMessagesProcessor answerProcessor,
                           TMessageTransmitter transmitter, ClientDetails recepient)
            throws IOException, StyxException, InterruptedException, TimeoutException {

        if (recepient == null) {
            throw new NullPointerException("recepient can't be null");
        }
        mRecepient = recepient;

        if (transmitter == null) {
            throw new NullPointerException("transmitter can't be null");
        }
        mTransmitter = transmitter;

        if (driver == null) {
            throw new NullPointerException("Channel driver can't be null");
        }
        setDriver(driver);

        if ( answerProcessor == null ) {
            throw new NullPointerException("answerProcessor can't be null");
        }
        mAnswerProcessor = answerProcessor;
        driver.setRMessageHandler(mAnswerProcessor);

        if (credentials == null) {
            throw new NullPointerException("Credentials can't be null");
        }
        mCredentials = credentials;
        mMountPoint = "/";
        sendVersionMessage();
        isConnected = driver.isConnected();

        return driver.isConnected();
    }

    /**
     * Connect to server with specified parameters
     *
     * @param driver channel driver
     * @return true if connected
     * @throws IOException
     * @throws StyxException
     * @throws TimeoutException
     */
    public boolean connect(IChannelDriver driver)
            throws IOException, StyxException, InterruptedException, TimeoutException {
        return connect(driver, mCredentials);
    }

    /**
     * Connect to server with specified parameters
     *
     * @return true if connection was success
     * @throws IOException
     * @throws StyxException
     * @throws InterruptedException
     * @throws TimeoutException
     */
    public boolean connect()
            throws IOException, StyxException, InterruptedException, TimeoutException {
        return connect(mDriver, mCredentials);
    }

    public StyxFile getRoot() throws StyxException, InterruptedException, TimeoutException, IOException {
        if (mRoot == null) {
            mRoot = new StyxFile(this, mFID);
        }
        return mRoot;
    }

    public IMessageTransmitter getMessenger() {
        return mTransmitter;
    }

    protected int getIOBufSize() {
        return DEFAULT_IO_SIZE;
    }

    public Credentials getCredentials() {
        return mCredentials;
    }

    public String getMountPoint() {
        return mMountPoint;
    }

    public long getRootFID() {
        return mFID;
    }

    public StyxQID getQID() {
        return mQID;
    }

    public long getAuthFID() {
        return mAuthFID;
    }

    public StyxQID getAuthQID() {
        return mAuthQID;
    }

    @Override
    public ConnectionDetails getConnectionDetails() {
        return mDetails;
    }

    public void sendVersionMessage()
            throws InterruptedException, StyxException, IOException, TimeoutException {
        // release attached FID
        if (mFID != StyxMessage.NOFID) {
            try {
                final StyxTMessageFID tClunk = new StyxTMessageFID(MessageType.Tclunk, MessageType.Rclunk, mFID);
                mTransmitter.sendMessage(tClunk, mRecepient);
            } catch (Exception e) {
                throw new IOException(e);
            }
            mFID = StyxMessage.NOFID;
        }

        StyxTVersionMessage tVersion = new StyxTVersionMessage(mDetails.getIOUnit(), getProtocol());
        mTransmitter.sendMessage(tVersion, mRecepient);

        StyxMessage rMessage = tVersion.waitForAnswer(mTimeout);
        StyxRVersionMessage rVersion = (StyxRVersionMessage) rMessage;
        if (rVersion.getMaxPacketSize() < mDetails.getIOUnit()) {
            mDetails = new ConnectionDetails(getProtocol(), (int) rVersion.getMaxPacketSize());
        }
        mRecepient.getPolls().getFIDPoll().clean();
        if ((mCredentials.getUserName() != null) && (mCredentials.getPassword() != null)) {
            sendAuthMessage();
        } else {
            sendAttachMessage();
        }
    }

    private void sendAuthMessage()
            throws InterruptedException, StyxException, IOException, TimeoutException {
        mAuthFID = mRecepient.getPolls().getFIDPoll().getFreeItem();

        StyxTAuthMessage tAuth = new StyxTAuthMessage(mAuthFID);
        tAuth.setUserName(getCredentials().getUserName());
        tAuth.setMountPoint(getMountPoint());
        mTransmitter.sendMessage(tAuth, mRecepient);

        StyxMessage rMessage = tAuth.waitForAnswer(mTimeout);
        StyxRAuthMessage rAuth = (StyxRAuthMessage) rMessage;
        mAuthQID = rAuth.getQID();

        // TODO uncomment later
        //        StyxOutputStream output = new StyxOutputStream((new StyxFile(this,
        //                ((StyxTAuthMessage)tMessage).getAuthFID())).openForWrite());
        //        output.writeString(getPassword());
        //        output.flush();

        sendAttachMessage();
    }

    private void sendAttachMessage()
            throws InterruptedException, StyxException, TimeoutException, IOException {
        mFID = mRecepient.getPolls().getFIDPoll().getFreeItem();
        StyxTAttachMessage tAttach = new StyxTAttachMessage(getRootFID(), getAuthFID(),
                getCredentials().getUserName(),
                getMountPoint());
        mTransmitter.sendMessage(tAttach, mRecepient);

        StyxMessage rMessage = tAttach.waitForAnswer(mTimeout);
        StyxRAttachMessage rAttach = (StyxRAttachMessage) rMessage;
        mQID = rAttach.getQID();
        setAttached(true);
    }

    @Override
    public void close() throws IOException {
        if (shouldCloseAnswerProcessor && mAnswerProcessor != null) {
            mAnswerProcessor.close();
            mAnswerProcessor = null;
        }
        if (mTransmitter != null) {
            mTransmitter.close();
            mTransmitter = null;
        }
        if (isAutoStartDriver && mDriver != null) {
            mDriver.close();
            mDriver = null;
        }
    }

    public String getProtocol() {
        return PROTOCOL;
    }

    //-------------------------------------------------------------------------------------
    // Getters
    //-------------------------------------------------------------------------------------
    public int getTimeout() {
        return mTimeout;
    }

    public boolean isAttached() {
        return isAttached;
    }

    //-------------------------------------------------------------------------------------
    // Setters
    //-------------------------------------------------------------------------------------
    public void setTimeout(int mTimeout) {
        this.mTimeout = mTimeout;
    }

    public void setAttached(boolean isAttached) {
        this.isAttached = isAttached;
    }

    public boolean isConnected() {
        return mDriver.isConnected();
    }

    private TMessageTransmitter.Listener mTransmitterListener = new TMessageTransmitter.Listener() {
        @Override
        public void onSocketDisconnected() {
            isConnected = false;
        }

        @Override
        public void onTrashReceived() {
            //something goes wrong, we should restart protocol
            setAttached(false);
            try {
                sendVersionMessage();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    };

    @Override
    public ClientDetails getRecepient() {
        return mRecepient;
    }

    protected void setDriver(IChannelDriver driver) {
        mDriver = driver;
    }
}

File: native/V2StyxLib/src/Connection.cpp
#include "Connection.h"
#include "types.h"
#include "messages/base/StyxTMessageFID.h"
#include "messages/StyxTVersionMessage.h"
#include "messages/StyxTAuthMessage.h"
#include "messages/StyxRAuthMessage.h"
#include "messages/StyxRVersionMessage.h"
#include "messages/StyxTAttachMessage.h"
#include "messages/StyxRAttachMessage.h"
#include "stdio.h"

const size_t Connection::DEFAULT_TIMEOUT = 10000;
const size_t Connection::DEFAULT_IOUNIT = 8192;

const StyxString Connection::PROTOCOL = StyxString("9P2000");

Connection::Connection(Credentials credentials,
				  IChannelDriver* driver,
				  RMessagesProcessor* answerProcessor,
				  TMessageTransmitter transmitter,
				  ClientDetails* recepient)
	: mTimeout(DEFAULT_TIMEOUT), mAuthQID(StyxQID::EMPTY), 
	mQID(StyxQID::EMPTY), mSyncObject(mTimeout), 
	mCredentials(credentials), mDriver(driver), mTransmitter(transmitter), isAutoStartDriver(false)  {
	if (mDriver == NULL) {
		throw StyxException("Channel driver can't be null");
	}
	printf("A20 %p\n", mDriver);
	mTransmitter.setListener(this);
	mAuthFID = StyxMessage::NOFID;
	mFID = StyxMessage::NOFID;
    shouldCloseAnswerProcessor = false;

	mRecepient = recepient;
	mAnswerProcessor = answerProcessor;
	mDetails = ConnectionDetails(getProtocol(), getIOBufSize());
	isConnectedFlag = false;
	mMountPoint = "/";
}

Connection::~Connection() {
}

void Connection::init(RMessagesProcessor *answerProcessor,
        TMessageTransmitter *transmitter,
        ClientDetails *recepient) {

}

void Connection::sendVersionMessage() throw(StyxException) {
    // release attached FID
    if (mFID != StyxMessage::NOFID) {
        try {
            StyxTMessageFID tClunk(Tclunk, Rclunk, mFID);
            mTransmitter.sendMessage(&tClunk, mRecepient);
        } catch (std::exception e) {
            throw e;
        }
        mFID = StyxMessage::NOFID;
    }

    StyxTVersionMessage tVersion(mDetails.getIOUnit(), getProtocol());
	StyxMessage* rMessage = mTransmitter.sendMessageAndWaitAnswer(&tVersion, mRecepient, &mSyncObject);
	StyxRVersionMessage* rVersion = (StyxRVersionMessage*) rMessage;
    if (rVersion->getMaxPacketSize() < mDetails.getIOUnit()) {
        mDetails = ConnectionDetails(getProtocol(), rVersion->getMaxPacketSize());
    }
    mRecepient->getFIDPoll()->clean();
    if ((!mCredentials.getUserName().empty()) && (!mCredentials.getPassword().empty())) {
        sendAuthMessage();
    } else {
        sendAttachMessage();
    }
}

size_t Connection::getIOBufSize() {
	return DEFAULT_IOUNIT;
}

void Connection::sendAuthMessage() throw(StyxException) {
	mAuthFID = mRecepient->getFIDPoll()->getFreeItem();

	StyxTAuthMessage tAuth(mAuthFID);
	tAuth.setUserName(mCredentials.getUserName());
	tAuth.setMountPoint(mMountPoint);
	StyxMessage* rMessage = mTransmitter.sendMessageAndWaitAnswer(&tAuth, mRecepient, &mSyncObject);
	StyxRAuthMessage* rAuth = (StyxRAuthMessage*) rMessage;
	mAuthQID = rAuth->getQID();

	// TODO uncomment later
	//        StyxOutputStream output = new StyxOutputStream((new StyxFile(this,
	//                ((StyxTAuthMessage)tMessage).getAuthFID())).openForWrite());
	//        output.writeString(getPassword());
	//        output.flush();

	sendAttachMessage();
}

void Connection::sendAttachMessage() throw(StyxException) {
	mFID = mRecepient->getFIDPoll()->getFreeItem();
	StyxTAttachMessage tAttach(getRootFID(), mAuthFID,
			mCredentials.getUserName(),
			mMountPoint);
	StyxMessage* rMessage = mTransmitter.sendMessageAndWaitAnswer(&tAttach, mRecepient, &mSyncObject);
	StyxRAttachMessage* rAttach = (StyxRAttachMessage*) rMessage;
	mQID = rAttach->getQID();
	isAttached = true;
}

StyxFID Connection::getRootFID() {
	return mFID;
}

size_t Connection::getTimeout() {
	return mTimeout;
}

IVirtualStyxFile* Connection::getRoot() {
	return mRoot;
}

ConnectionDetails Connection::getConnectionDetails() {
	return mDetails;
}

ClientDetails *Connection::getRecepient() {
	return mRecepient;
}

void Connection::close() throw(StyxException) {
	if (shouldCloseAnswerProcessor && mAnswerProcessor != NULL) {
		mAnswerProcessor->close();
		delete mAnswerProcessor;
		mAnswerProcessor = NULL;
	}
	mTransmitter.close();
	if (isAutoStartDriver) {
		mDriver->close();
		mDriver = NULL;
	}
}

StyxString Connection::getProtocol() {
	return PROTOCOL;
}

Credentials Connection::getCredentials() {
	return mCredentials;
}

StyxQID Connection::getQID() {
	return mQID;
}

bool Connection::isConnected() {
	return mDriver->isConnected();
}

IMessageTransmitter *Connection::getTransmitter() {
	return &mTransmitter;
}

StyxString Connection::getMountPoint() {
	return mMountPoint;
}

bool Connection::connect() throw(StyxException) {
#warning move to builder
	if (mAnswerProcessor == NULL) {
		mAnswerProcessor = new RMessagesProcessor("RH" + mDriver->toString());
		shouldCloseAnswerProcessor = true;
	}
	if (!mDriver->isStarted()) {
		mDriver->setRMessageHandler(mAnswerProcessor);
#warning this temp fix. TMessages shoul be handled in other way.
		mDriver->setTMessageHandler(mAnswerProcessor); 
		mDriver->start(getIOBufSize());
		isAutoStartDriver = true;
	}

	if (mRecepient == NULL) {
		// get first client from driver
		mRecepient = mDriver->getClients().front();
	}
	if (mRecepient == NULL) {
        throw StyxException("Recipient can't be null");
    }

	printf("A10 %p\n", mDriver);
    sendVersionMessage();
    return mDriver->isConnected();
}

void Connection::setAttached(bool isAttached) {
	this->isAttached = isAttached;
}


void Connection::onTrashReceived(TMessageTransmitter *caller) {
    //something goes wrong, we should restart protocol
    setAttached(false);
    try {
        sendVersionMessage();
    } catch (StyxException e) {
        e.printStackTrace();
    }
}

void Connection::onChannelDisconnected(TMessageTransmitter *caller) {
	isConnectedFlag = false;
}

Connection::Builder::Builder() : mCredentials("", ""), mDriver(NULL), mAnswerProcessor(NULL),
	mTransmitter(), mClientDetails(NULL) 
{
}

Connection::Builder::~Builder() 
{
}

Connection* Connection::Builder::build() {
	return new Connection(mCredentials, mDriver, mAnswerProcessor, mTransmitter, mClientDetails);
}

Connection::Builder* Connection::Builder::setDriver(IChannelDriver* driver) {
	mDriver = driver;
	return this;
}